iris-mpc-cpu/src/network/grpc.rs

```diff
@@ -29,94 +28,334 @@ use tonic::{
-    ) {
-        self.streams
-            .write()
-            .await
-            .insert((session_id, receiver_id), Arc::new(stream));
+    fn add_session_stream(&mut self, session_id: SessionId, receiver_id: Identity, stream: Sender) {
+        self.streams.insert((session_id, receiver_id), stream);
     }
 
-    async fn get_stream(
-        &self,
-        session_id: SessionId,
-        receiver_id: Identity,
-    ) -> eyre::Result<Arc<Sender>> {
+    fn get_stream(&self, session_id: SessionId, receiver_id: Identity) -> eyre::Result<Sender> {
         self.streams
-            .read()
-            .await
             .get(&(session_id, receiver_id.clone()))
             .ok_or(eyre!(
                 "Streams for session {session_id:?} and receiver {receiver_id:?} not found"
             ))
-            .map(Arc::clone)
+            .cloned()
     }
 
-    async fn count_receivers(&self, session_id: SessionId) -> usize {
+    fn count_receivers(&self, session_id: SessionId) -> usize {
         self.streams
-            .read()
-            .await
             .iter()
             .filter(|((sid, _), _)| *sid == session_id)
             .count()
     }
 }
 
+struct SendTask {
+    value: Vec<u8>,
+    receiver: Identity,
+    session_id: SessionId,
+}
+
+struct ReceiveTask {
+    sender: Identity,
+    session_id: SessionId,
+}
+
+struct StartMessageStreamTask {
+    sender: Identity,
+    session_id: SessionId,
+    stream: UnboundedReceiver<SendRequest>,
+}
+
+struct ConnectToPartyTask {
+    party_id: Identity,
+    address: String,
+}
+
+enum GrpcTask {
+    ConnectToParty(ConnectToPartyTask),
+    CreateSession(SessionId),
+    StartMessageStream(StartMessageStreamTask),
+    IsSessionReady(SessionId),
+    Send(SendTask),
+    Receive(ReceiveTask),
+}
+
+enum MessageResult {
+    Empty,
+    IsSessionReady(bool),
+    ReceivedBytes(Vec<u8>),
+}
+
+struct MessageJob {
+    task: GrpcTask,
+    return_channel: oneshot::Sender<eyre::Result<MessageResult>>,
+}
+
+// Concurrency handler for networking operations
+#[derive(Clone)]
+pub struct GrpcHandle {
+    job_queue: mpsc::Sender<MessageJob>,
+    party_id: Identity,
+}
+
+impl GrpcHandle {
+    pub async fn new(mut grpc: GrpcNetworking) -> eyre::Result<Self> {
+        let party_id = grpc.party_id.clone();
+        let (tx, mut rx) = tokio::sync::mpsc::channel::<MessageJob>(1);
+
+        // Loop to handle incoming tasks from job queue
+        tokio::spawn(async move {
+            while let Some(job) = rx.recv().await {
+                match job.task {
+                    GrpcTask::Send(task) => {
+                        let job_result = grpc
+                            .send(task.value, &task.receiver, &task.session_id)
+                            .map(|_| MessageResult::Empty);
+                        let _ = job.return_channel.send(job_result);
+                    }
+                    GrpcTask::Receive(task) => {
+                        let mut grpc = grpc.clone();
```

Comment:
That is a really expensive clone. This could lookup the message queue first, and clone only an `Arc` to send into the `spawn`.

The thing being cloned could be one of:

- Lookup the session and pass `Arc<MessageQueueStore>`. Need an Arc but no need for Mutex once it is constructed,
- or lookup up the session and sender too and send `Arc<Mutex<UnboundedReceiver<SendRequest>>>`,
- or for the brave, take the lock and send [`OwnedMutexGuard<UnboundedReceiver<SendRequest>>`](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.lock_owned)

Reply:
I think option 2 is the best, WDYT?

Reply:
Yep, the second options works nicely.

Reply:
Done